xml/System/Index.xml

```diff
@@ -251,10 +274,24 @@
         <Parameter Name="length" Type="System.Int32" />
       </Parameters>
       <Docs>
-        <param name="length">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="length">The length of the collection that the Index will be used with. Must be a positive value.</param>
+        <summary>Calculate the offset from the start using the giving collection length.</summary>
+        <returns>The offset.</returns>
+        <remarks>
+          <format type="text/markdown">
+            <![CDATA[
+
+### Remarks
+
+For performance reasons, we don't validate the input length parameter and the returned offset value against negative values.
+
+We also don't validate if the returned offset is greater than the input length.
+
+It is expected that Index will always be used with collections have non negative length/count. If the returned offset is negative and then used to index, a collection will throw range exception which will be the same effect as the validation.
```

Comment:
```suggestion
If the returned offset is negative and is then used as an index into a collection, the operation throws an <xref:System.ArgumentOutOfRangeException>. 
```
I don't understand this original comment. Is it even possible to have a collection or an array with a negative count or length? It's possible for an array to have an arbitrary starting index, but that doesn't affect its length.

Reply:
@tarekgh can you also help explain this comment a bit more?

Reply:
Here is some clarification:

before I explain the comments, I want to clarify what is the purpose of this method

Index can be created not tied to any collection (think about it as int value). The index can be from the beginning or from the end. The callers of GetOffset is going to path collection length to calculate the the offset which the Index point to. If the index from start, then we just return whatever stored offset. If the index from end, we'll use the input length to calculate the offset. Here is some examples: 

imagine you have index from start with value 1, and then calling GetOffset(10), this will return 1 because this is the offset from the beginning of the collection).
If you have Index with value 1 but from the end, and calling GetOffset(10) this will return 9. which is the offset from the beginning of  collection.

Now let's explain the remark written there:

#1
What happen if someone pass negative length value (i.e. GetOffset(-10)), this should be invalid. For performance reason we don't validate the input length and we don't even fail but we'll return some invalid result too. It is the caller responsibility to pass correct logical values. 

 #2
If you have Index with value 10 and then calling GetOffset(4), this also is invalid input because the used index is outside the range of the passed collection length. again for performance reason we don't validate that and we'll return invalid value. It is the caller responsibility to pass correct logical values.

#3
Is like #2, imagine you have a collection with length 4 and then have Index with value 10, calling GetOffset(4) and Index is from start, then this will return 10 and when you use 10 to access the collection you'll get out of range exception. 
If the index is from the end, GetOffset(4) will negative value (-6) and when you use this value to access the collection will get OutOfRangeException. 

We are trying to say even we are not validating the input and the result of GetOffset, in most common cases when operating over a collection, will get OutOfRangeException if used invalid input to GetOffset. 

Let me know if this is still not clear.


---
In reply to: [305165870](https://github.com/dotnet/dotnet-api-docs/pull/2729#discussion_r305165870) [](ancestors = 305165870)

Reply:
Ok I understand now, thanks for the explanation. I'll reword it a bit:

```suggestion
Collections are not expected to have a non-negative length/count. If the returned offset is negative and then used to index the collection, the collection will throw <xref:System.ArgumentOutOfRangeException>, which will have the same effect as validation.
```